internal/controller/snapshot/snapshot_adapter.go

```diff
@@ -433,6 +435,73 @@ func (a *Adapter) EnsureAllReleasesExist() (controller.OperationResult, error) {
 	return controller.ContinueProcessing()
 }
 
+// EnsureOverrideSnapshotValid is an operation that ensure the manually created override snapshot have valid
+// digest and git source in snapshotComponents, mark it as invalid if it is invalid
+func (a *Adapter) EnsureOverrideSnapshotValid() (controller.OperationResult, error) {
+	if !gitops.IsOverrideSnapshot(a.snapshot) {
+		a.logger.Info("The snapshot was not override snapshot, skipping")
+		return controller.ContinueProcessing()
+	}
+
+	if gitops.IsSnapshotMarkedAsInvalid(a.snapshot) {
+		a.logger.Info("The override snapshot has been marked as invalid, skipping")
+		return controller.ContinueProcessing()
+	}
+
+	var err error
+	if !controllerutil.HasControllerReference(a.snapshot) {
+		a.snapshot, err = gitops.SetOwnerReference(a.context, a.client, a.snapshot, a.application)
+		if err != nil {
+			a.logger.Error(err, fmt.Sprintf("Failed to set owner reference for snapshot %s/%s", a.snapshot.Namespace, a.snapshot.Name))
+			return controller.RequeueWithError(err)
+		}
+		a.logger.Info("Owner reference has been set to snapshot")
+	}
+
+	// validate all snapshotComponents' containerImages/source in snapshot, make all errors joined
+	var errsForSnapshot error
+	for _, snapshotComponent := range a.snapshot.Spec.Components {
+		snapshotComponent := snapshotComponent //G601
+		_, err := a.loader.GetComponent(a.context, a.client, snapshotComponent.Name, a.snapshot.Namespace)
+		if err != nil {
+			a.logger.Error(err, "Failed to get component from applicaion", "application.Name", a.application.Name, "component.Name", snapshotComponent.Name)
+			_, loaderError := h.HandleLoaderError(a.logger, err, snapshotComponent.Name, a.application.Name)
+			if loaderError != nil {
+				return controller.RequeueWithError(loaderError)
+			} else {
+				errsForSnapshot = errors.Join(errsForSnapshot, fmt.Errorf("snapshotComponent %s defined in snapshot %s doesn't exist under application %s/%s", snapshotComponent.Name, a.snapshot.Name, a.application.Namespace, a.application.Name))
+				continue
+			}
+		}
+
+		err = gitops.ValidateImageDigest(snapshotComponent.ContainerImage)
+		if err != nil {
+			a.logger.Error(err, "containerImage in snapshotComponent has no valid digest", "snapshotComponent.Name", snapshotComponent.Name, "snapshotComponent.ContainerImage", snapshotComponent.ContainerImage)
+			errsForSnapshot = errors.Join(errsForSnapshot, err)
+			continue
+		}
+
+		if !gitops.HaveGitSource(snapshotComponent) {
+			a.logger.Error(err, "snapshotComponent has no git url/revision", "snapshotComponent.Name", snapshotComponent.Name, "snapshotComponent.ContainerImage", snapshotComponent.ContainerImage)
+			errsForSnapshot = errors.Join(errsForSnapshot, err)
+			continue
+		}
+	}
+
+	if errsForSnapshot != nil {
+		a.logger.Error(errsForSnapshot, "mark the override snapshot as invalid due to invalid containerImage or git source")
```

Comment:
This could be misleading as the snapshot could be invalid due to absence of snapshotComponent, so we could either mention all the possible reasons like this:
```suggestion
		a.logger.Error(errsForSnapshot, "mark the override snapshot as invalid due to non-existent components or git source or due invalid containerImage defined within it")
```
or mention none of them since they will end up printed by the `errsForSnapshot` field:
```suggestion
		a.logger.Error(errsForSnapshot, "mark the override snapshot as invalid")
```
wdyt?

Reply:
Thanks. Let's change to `Snapshot has been marked as invalid due to invalid snapshotComponent`

Reply:
Nice, that makes sense to me.